### PR TITLE
RLookup refactor - APPLY/FILTER refactor - [MOD-5085]

### PR DESCRIPTION
**Pull request by GuyAv46**

### CODE REVIEW

File: src/rlookup.c [hunk 1]
```diff
@@ -24,17 +24,50 @@ static RLookupKey *createNewKey(RLookup *lookup, const char *name, size_t name_l
...
 }
 
+// Allocate a new RLookupKey and add it to the RLookup table.
+static RLookupKey *overrideKey(RLookup *lk, RLookupKey *old, uint32_t flags) {
+  RLookupKey *new = rm_calloc(1, sizeof(*new));
+
+  // Copy the old key to the new one
+  new->name = old->name; // taking ownership of the name
+  new->name_len = old->name_len;
+  new->path = new->name; // keeping the default of path = name
```
Comment (DvirDukhan) [inline src/rlookup.c:46]:
```suggestion
  new->path = new->name; // keeping the initial setting of path = name, path resultion will happen later
```

File: src/rlookup.c [hunk 2]
```diff
@@ -24,17 +24,50 @@ static RLookupKey *createNewKey(RLookup *lookup, const char *name, size_t name_l
...
+  new->dstidx = old->dstidx;
+
+  // Set the new flags
+  new->flags = flags & ~RLOOKUP_TRANSIENT_FLAGS;
+  // If the old key was allocated, we take ownership of the name.
+  new->flags |= old->flags & RLOOKUP_F_NAMEALLOC;
+
+  // Make the old key inaccessible for new lookups
+  old->flags &= (old->path == old->name) ? ~RLOOKUP_F_NAMEALLOC : UINT32_MAX; // If the old key allocated the name and not the path, we take ownership of the allocation
```
Comment (DvirDukhan) [inline src/rlookup.c:55]: either break into if
or
define new mask and call it (the mask value is UNIT32_MASK

File: src/rlookup.c [hunk 3]
```diff
@@ -24,17 +24,50 @@ static RLookupKey *createNewKey(RLookup *lookup, const char *name, size_t name_l
...
+  old->name = NULL;
+  old->name_len = -1; // 0 is a valid length if the user provided an empty string as a name
```
Comment (DvirDukhan) [inline src/rlookup.c:57]: add a comment about why this is safe

File: coord/src/dist_plan.cpp [hunk 4]
```diff
@@ -331,6 +331,9 @@ static int distributeAvg(ReducerDistCtx *rdctx, QueryError *status) {
   char *expr = rm_strdup(ss.c_str());
   PLN_MapFilterStep *applyStep = PLNMapFilterStep_New(expr, PLN_T_APPLY);
   applyStep->shouldFreeRaw = 1;
+  applyStep->noOverride = 1; // Don't override the alias. Usually we do, but in this case we don't because reducers
+                             // are not allowed to override aliases
+  applyStep->readHidden = 1; // Keep the hidden reducers hidden
```
Comment (oshadmi) [inline coord/src/dist_plan.cpp:336]: Remove readHidden ?

File: src/aggregate/expr/expression.c [hunk 5]
```diff
@@ -265,7 +265,7 @@ int ExprAST_GetLookupKeys(RSExpr *expr, RLookup *lookup, QueryError *err) {
 
   switch (expr->t) {
     case RSExpr_Property:
-      expr->property.lookupObj = RLookup_GetKey_TEMP(lookup, expr->property.key, RLOOKUP_F_NOFLAGS);
+      expr->property.lookupObj = RLookup_GetKey(lookup, expr->property.key, RLOOKUP_M_READ, RLOOKUP_F_NOFLAGS);
       if (!expr->property.lookupObj) {
         QueryError_SetErrorFmt(err, QUERY_ENOPROPKEY, "Property `%s` not loaded in pipeline",
```
Comment (oshadmi) [inline src/aggregate/expr/expression.c:270]: Not necessarily "not loaded"?

**Review by oshadmi — APPROVED** [review]

**Pull request merged**